### PR TITLE
3.0 alpha

### PR DESCRIPTION
**Pull request by hegdeashwin**
* Migrated from JShint to ESLint
* Migration setup from ES5 to ES6 (Used Babel for safe transpiling support)
* Removes "use strict" as in ES6 modules, are always have strict mode semantics, the directives are unnecessary.
* Fixes some const and let
* Also removed the dependence of Grunt. Now, all the thing should work via npm scripts

image

### CODE REVIEW

**Comment by hegdeashwin** [comment]
Below are the command which are working:

`npm run lint` - will lint `lib`, `examples` and `test` directories. This also includes ESLint `--fix` autofix feature.
`npm run test` - will perform lint and then execute test cases via Mocha.
`npm run test:default` - will perform lint and the execute test cases via Mocha with default option `--timeout 30000` and `--reporter spec`.

File: examples/chain.js [hunk 1]
```diff
@@ -1,9 +1,9 @@
...
-  "driver": {
-    "browser": "phantomjs"
-  },
-  'data': {
+const Nemo = require('../');
+const config = {
+  'driver': {
+    'browser': 'phantomjs'
+  }
+  ,'data': {
```
Comment (grawk) [inline examples/chain.js]: What is the rationale for making the change to a "leading" comma in some of these cases?

Comment (skratchdot) [inline examples/chain.js]: I was going to make the same comment.  It's a small thing, but it jumped out at me as unnecessary.

File: index.js [hunk 2]
```diff
@@ -12,16 +12,15 @@
  │   See the License for the specific language governing permissions and       │
  │   limitations under the License.                                            │
  \*───────────────────────────────────────────────────────────────────────────*/
-'use strict';
```
Comment (skratchdot) [inline index.js]: I vaguely remember running into a const/let error in node v4 when removing a use strict, but I can't seem to reproduce it.  As long as things are still working in v4, this seems fine to me.

File: package.json [hunk 3]
```diff
@@ -16,12 +18,17 @@
   },
   "main": "index.js",
   "devDependencies": {
+    "babel-core": "^6.24.1",
+    "babel-preset-env": "^1.4.0",
     "chai": "^3.5.0",
+    "eslint": "^3.19.0",
+    "eslint-plugin-json": "^1.2.0",
     "grunt": "^1.0.1",
```
Comment (skratchdot) [inline package.json]: if we are actually removing the gruntfile.js, then there's no need for these 3 grunt libs here (unless there's something I'm missing / not thinking about).

File: examples/chain.js (hunk 1 above)
Comment (hegdeashwin) [inline examples/chain.js]: @grawk , @skratchdot , actually '--fix' parameter auto-fix it. Although this looks small thing but let me observe this behavior ... and will update soon on this.

File: index.js (hunk 2 above)
Comment (hegdeashwin) [inline index.js]: @skratchdot : I tested it on v6.10.2 & v7.9.0 its working ... will check on v4 ranges and update soon

File: package.json (hunk 3 above)
Comment (hegdeashwin) [inline package.json]: Yes! I found my package.json was not getting updated even after the npm uninstall command, that the reason they are still there ... no problem ... will update that too

File: index.js (hunk 2 above)
Comment (skratchdot) [inline index.js]: @hegdeashwin - actually, this is against the 3.0-alpha branch, so we only need to support node v6. I think we are good here. Thanks for the response.

File: package.json [hunk 4]
```diff
@@ -16,12 +18,17 @@
   },
   "main": "index.js",
   "devDependencies": {
+    "babel-core": "^6.24.1",
+    "babel-preset-env": "^1.4.0",
     "chai": "^3.5.0",
+    "eslint": "^3.19.0",
+    "eslint-plugin-json": "^1.2.0",
```
Comment (skratchdot) [inline package.json]: It doesn't look like you are actually configuring `.eslintrc` to use `eslint-plugin-json`.  Do we need that plugin if we aren't using it?

File: package.json [hunk 5]
```diff
@@ -16,12 +18,17 @@
...
     "grunt": "^1.0.1",
     "grunt-contrib-jshint": "^1.0.0",
     "grunt-simple-mocha": "~0.4.0",
     "mocha": "^2.0.0",
-    "nconf": "^0.8.4"
+    "nconf": "^0.8.4",
+    "phantomjs": "^2.1.7"
```
Comment (skratchdot) [inline package.json]: We are adding `phantomjs` as a devDependency due to the `examples/chain.js` file, correct?

**Comment by manoj9788** [comment]
@grawk during our discussion in SeConf you said, you don't want transpiling using Babel, that's the reason on my PR I just started slowly with minimal es6 changes. - Just curious?

**Comment by grawk** [comment]
@manoj9788 afaict there is no actual usage of babel in this PR. I definitely don't want to be using it.

**Comment by manoj9788** [comment]
Ah! I went by Ashwin's description.. so it's not actually babel usage. Alright!

File: index.js (hunk 2 above)
Comment (hegdeashwin) [inline index.js]: 👍 Thank you for the quick response

File: package.json (hunk 5 above)
Comment (hegdeashwin) [inline package.json]: I felt the same, Also, try with Yarn but did not worked as per my expectation ... will exclude it from package.json

File: package.json [hunk 6]
```diff
@@ -16,12 +18,17 @@
   },
   "main": "index.js",
   "devDependencies": {
+    "babel-core": "^6.24.1",
```
Comment (hegdeashwin) [inline package.json]: 👍

File: package.json (hunk 4 above)
Comment (hegdeashwin) [inline package.json]: Fixes done in 29a3d5b7561c19f07f2fd5c0fb94900916a94bf0

File: package.json (hunk 6 above)
Comment (hegdeashwin) [inline package.json]: Fixes done in f0e0eb148d078e7e46a31fddb26c86960337dc36

File: package.json (hunk 5 above)
Comment (hegdeashwin) [inline package.json]: Fixes done in f0e0eb148d078e7e46a31fddb26c86960337dc36

File: examples/chain.js (hunk 1 above)
Comment (hegdeashwin) [inline examples/chain.js]: Fixes done in d554742f83922e56b3d8eb06343fe31267fae5b1 

Some useful eslint links:
http://eslint.org/docs/rules/comma-dangle
http://eslint.org/docs/rules/comma-spacing
http://eslint.org/docs/rules/comma-style

**Comment by hegdeashwin** [comment]
Also, grunt-cli global install in Travis is no more required. So updated the Travis YML configuration. Please review the commit 091fd0359bd62a7c99e8db8659d7ca98a0e7e182

File: test/plugin.js [hunk 7]
```diff
@@ -41,13 +39,13 @@ describe('@plugin@', function () {
...
-        "browser": "phantomjs"
+      'driver': {
+        'browser': 'phantomjs'
       },
       'plugins': {
         'crappy': {
-          'module': 'path:plugin/sample',
-          'arguments': ['crap plugin']
+          'module': 'path:plugin/sample'
+          ,'arguments': ['crap plugin']
```
Comment (grawk) [inline test/plugin.js]: remaining "leading" comma

File: examples/chain.js (hunk 1 above)
Comment (grawk) [inline examples/chain.js]: remaining "leading" comma

**Comment by grawk** [comment]
The build looks fine. Just need a couple of these auto-fixed leading commas to be repaired and should be good to go.

File: examples/chain.js (hunk 1 above)
Comment (hegdeashwin) [inline examples/chain.js]: Will need to check. Only fixed those `--fix` parameters impacted ... will update soon

File: test/plugin.js (hunk 7 above)
Comment (hegdeashwin) [inline test/plugin.js]: 👍  gotcha

Comment (hegdeashwin) [inline test/plugin.js]: Found 2 uncaught files. Fixes done in 332de79b7ffba1d4d233c3cb7e62436c0ce41349

File: examples/chain.js (hunk 1 above)
Comment (hegdeashwin) [inline examples/chain.js]: Found 2 uncaught files. Fixes done in 332de79b7ffba1d4d233c3cb7e62436c0ce41349

**Comment by skratchdot** [comment]
this lgtm

**Comment by grawk** [comment]
Fine, I guess the kids these days don't like parens around their ternary conditions 😆 

LGTM!

File: lib/configure.js [hunk 8]
```diff
@@ -17,10 +15,10 @@ module.exports = function Configure(_basedir, _configOverride) {
   log('_basedir %s, _configOverride %o', _basedir, _configOverride);
   let basedir, configOverride;
   //settle arguments
-  basedir = (arguments.length && typeof arguments[0] === 'string') ? arguments[0] : process.env.nemoBaseDir || undefined;
-  configOverride = (!basedir && arguments.length && typeof arguments[0] === 'object') ? arguments[0] : undefined;
-  configOverride = (!configOverride && arguments.length && arguments[1] && typeof arguments[1] === 'object') ? arguments[1] : configOverride;
-  configOverride = (!configOverride) ? {} : configOverride;
+  basedir = arguments.length && typeof arguments[0] === 'string' ? arguments[0] : process.env.nemoBaseDir || undefined;
```
Comment (hegdeashwin) [inline lib/configure.js]: ESlint is auto fixing this as per the eslint:recommended rules. There 4 rules on ternary operators but non of them says about parentheses.

**Pull request merged**